### PR TITLE
Initial support for BG95-M6 modem [sc-110723]

### PR DESCRIPTION
**Pull request by tstellanova**
### Problem

BG95-M6 modem support missing

### Solution

- Allocate a new ncpId, `PLATFORM_NCP_QUECTEL_BG95_M6`, and make necessary changes for BG95-M6 modem
- Remove duplicate 650ms timing code for modemPowerOff()
- Remove incorrect 200ms timing code for modemPowerOn() [changed here.](https://github.com/particle-iot/device-os/pull/2110/files#diff-186ce5679947e07e5ce6c3d4fce9cdce5ebd61e4c315bcf0658db6106ea54299L1349-R1425)
- Remove AT+CMEE? polling during registration for Quectel devices. AT+CMEE? will always report +CMEE: 1 which is the factory default, so no reason to poll it continuously while waiting for registration.
- Remove AT+CEER for all devices not supporting it.  Only supported on EG91 Cat-1 devices.
- Speed up timeout wait times for modemPowerOn/modemPowerOff based on each modem specification, instead of using the very long time of EG91 for all.

### Steps to Test

- Validate with a Particle device built with the BG95-M6 modem, and with appropriated programmed ncpId,
using eg cellular test tools. 
- Run HIL tests with BG95-M1/EG91/BG96 devices

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [x] Run unit/integration/application tests on device
- [x] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 1]
```diff
@@ -1790,7 +1799,7 @@ int QuectelNcpClient::processEventsImpl() {
     SCOPE_GUARD({ regCheckTime_ = millis(); });
 
     // Check GPRS, LET, NB-IOT network registration status
-    if (ncpId() != PLATFORM_NCP_QUECTEL_BG95_M1) {
+    if (!isQuecLPWADevice()) {
```
Comment (technobly) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1802]: We need to be careful here and make sure that this works correctly.  I believe the intention of != M1, was to say these commands error on M1 specifically.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 2]
```diff
@@ -1878,7 +1887,7 @@ int QuectelNcpClient::modemPowerOn() {
     if (!modemPowerState()) {
         ncpPowerState(NcpPowerState::TRANSIENT_ON);
 
-        if (ncpId() != PLATFORM_NCP_QUECTEL_BG95_M1 && ncpId() != PLATFORM_NCP_QUECTEL_BG95_MF) {
+        if (!isQuecLPWADevice()) {
```
Comment (technobly) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1890]: We need to be careful here and make sure that this works correctly.  BG77 might fall into the other category.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 3]
```diff
@@ -1919,7 +1928,7 @@ int QuectelNcpClient::modemPowerOff() {
 
         LOG(TRACE, "Powering modem off");
 
-        if (ncpId() != PLATFORM_NCP_QUECTEL_BG95_M1 && ncpId() != PLATFORM_NCP_QUECTEL_BG95_MF) {
+        if (!isQuecLPWADevice()) {
```
Comment (technobly) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1931]: We need to be careful here and make sure that this works correctly. BG77 might fall into the other category.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 4]
```diff
@@ -1992,7 +2001,7 @@ int QuectelNcpClient::modemSoftPowerOff() {
 int QuectelNcpClient::modemHardReset(bool powerOff) {
     LOG(TRACE, "Hard resetting the modem");
 
-    if (ncpId() != PLATFORM_NCP_QUECTEL_BG95_M1 && ncpId() != PLATFORM_NCP_QUECTEL_BG95_MF) {
+    if (!isQuecLPWADevice()) {
```
Comment (technobly) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:2004]: We need to be careful here and make sure that this works correctly. BG77 might fall into the other category.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 5]
```diff
@@ -1882,26 +1905,26 @@ int QuectelNcpClient::modemPowerOn() {
...
+        hal_gpio_write(BGPWR, 0);
 
         // After power on the device, we can't assume the device is ready for operation:
-        // BG95: status pin ready requires >= 2.1s, uart ready requires >= 2.5s
+        // BG95/BG77: status pin ready requires >= 2.1s, uart ready requires >= 2.5s
         // BG96: status pin ready requires >= 4.8s, uart ready requires >= 4.9s
         // EG91: status pin ready requires >= 10s, uart ready requires >= 12s
-        if (waitModemPowerState(1, 15000)) {
+        // Wait 3 seconds longer than minimum spec
+        int ncp_id = ncpId();
```
Comment (avtolstoy) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1920]: Nit: It's fine to just `ncpId() ==` without temp variable, it'll be optimized just fine.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 6]
```diff
@@ -1793,14 +1809,21 @@ int QuectelNcpClient::processEventsImpl() {
     }
     SCOPE_GUARD({ regCheckTime_ = millis(); });
 
-    // Check GPRS, LET, NB-IOT network registration status
-    if (ncpId() != PLATFORM_NCP_QUECTEL_BG95_M1) {
-        CHECK_PARSER(parser_.execCommand("AT+CEER"));  // Seems to stall the modem on BG95-MF
-        CHECK_PARSER(parser_.execCommand("AT+CMEE?")); // Seems to stall the modem on BG95-MF
```
Comment (avtolstoy) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1799]: Let's leave this one in as well. It's another error code that is nice to get on supported modems.

**Review by avtolstoy — APPROVED** [review]

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp (hunk 6 above)
Comment (technobly) [reply hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1799]: AT+CMEE? ?  That's just the level of CMEE set which is always 1.

**Comment by jmwilson** [comment]
Connects to the cloud on Tracker M v0.0.3 with this branch's changes 💯

**Pull request merged**